ducktape/tests/runner_client.py

```diff
@@ -0,0 +1,281 @@
+        self.test_context = self.collect_test_context(**self.test_metadata)
+
+        self.send(self.request.running())
+        if len(self.cluster) != self.cluster.num_available_nodes():
+            # Sanity check - are we leaking cluster nodes?
+            raise RuntimeError(
+                "Expected all nodes to be available. Instead, %d of %d are available" %
+                (self.cluster.num_available_nodes(), len(self.cluster)))
+
+        if self.test_context.ignore:
+            # Skip running this test, but keep track of the fact that we ignored it
+            result = TestResult(self.test_context,
+                                self.session_context,
+                                test_status=IGNORE,
+                                start_time=time.time(),
+                                stop_time=time.time())
+            # Tell the server we are finished
+            self.send(self.request.finished(result=result))
+            return
+
+        # Results from this test, as well as logs will be dumped here
+        mkdir_p(self.test_context.results_dir)
+
+        start_time = -1
+        stop_time = -1
+        test_status = PASS
+        summary = ""
+        data = None
+
+        try:
+            # Instantiate test
+            self.test = self.test_context.cls(self.test_context)
+
+            self.log(logging.DEBUG, "Checking if there are enough nodes...")
+            if self.test.min_cluster_size() > len(self.cluster):
+                raise RuntimeError(
+                    "There are not enough nodes available in the cluster to run this test. "
+                    "Cluster size: %d, Need at least: %d. Services currently registered: %s" %
+                    (len(self.cluster), self.test.min_cluster_size(), self.test_context.services))
+
+            # Run the test unit
+            start_time = time.time()
+            self.setup_test()
+
+            data = self.run_test()
+            test_status = PASS
+            self.log(logging.INFO, "PASS")
+
+        except BaseException as e:
+            err_trace = str(e.message) + "\n" + traceback.format_exc(limit=16)
+            self.log(logging.INFO, "FAIL: " + err_trace)
+
+            test_status = FAIL
+            summary += err_trace
+
+        finally:
+            self.teardown_test(teardown_services=not self.session_context.no_teardown)
+            stop_time = time.time()
+
+            result = TestResult(
+                self.test_context,
+                self.session_context,
+                test_status,
+                summary,
+                data,
+                start_time,
+                stop_time)
+
+            self.log(logging.INFO, "Summary: %s" % str(result.summary))
+            self.log(logging.INFO, "Data: %s" % str(result.data))
+
+            test_reporter = SingleResultFileReporter(result)
+            test_reporter.report()
+
+        # Tell the server we are finished
```

Comment:
yep it'll cause the parent process to hang :)
I plan to add a "liveness" check ("is the subprocess pid still present?") in the event loop. Even with such a check, the parent process will still hang if the test itself hangs and the subprocess never exits.

It's interesting that even to handle the non-parallel case, a liveness check on the subprocesses can help handle certain failures cleanly.